File: reports/labels.go
package reports

import (
	"fmt"
	"strconv"

	"github.com/DavidArthurCole/EggLedger/ei"
	"github.com/DavidArthurCole/EggLedger/ledgerdata"
)

var rarityNames = []string{"Common", "Rare", "Epic", "Legendary"}
var tierNames = []string{"T1", "T2", "T3", "T4"}

// artifactDisplayName returns the human-readable display name for an artifact
// enum value by looking it up in ledgerdata.Config.ArtifactTargets. Falls back
// to CasedName() if no match is found.
func artifactDisplayName(v int32) string {
	protoName := ei.ArtifactSpec_Name(v).String()
	for _, t := range ledgerdata.Config.ArtifactTargets {
		if t.Name == protoName {
			return t.DisplayName
		}
	}
	return ei.ArtifactSpec_Name(v).CasedName()
}

// numericGroupBys lists every GROUP BY dimension whose raw SQL value is an integer.
var numericGroupBys = map[string]bool{
	"ship_type":      true,
	"duration_type":  true,
	"level":          true,
	"mission_type":   true,
	"rarity":         true,
	"tier":           true,
	"artifact_name":  true,
	"mission_target": true,
}

// LabelSortLess returns true when raw value rawA should sort before rawB for the
// given groupBy dimension. Numeric dimensions compare by integer value; all others
// compare lexicographically.
func LabelSortLess(groupBy, rawA, rawB string) bool {
	if numericGroupBys[groupBy] {
		a, errA := strconv.ParseInt(rawA, 10, 64)
		b, errB := strconv.ParseInt(rawB, 10, 64)
		if errA == nil && errB == nil {
			return a < b
		}
	}
	return rawA < rawB
}

// FormatLabel converts a raw SQL GROUP BY value to a display string.
func FormatLabel(groupBy string, rawVal string) string {
	parseInt := func() int {
		v, _ := strconv.Atoi(rawVal)
		return v
	}
	switch groupBy {
	case "ship_type":
		return ei.MissionInfo_Spaceship(parseInt()).Name()
	case "duration_type":
		return ei.MissionInfo_DurationType(parseInt()).Display()
	case "level":
		return fmt.Sprintf("Level %s", rawVal)
	case "mission_type":
		return ei.MissionInfo_MissionType(parseInt()).Display()
	case "mission_target":
		v := parseInt()
		if v < 0 {
			return "None (Pre 1.27)"
		}
		if v == 0 {
			return "Untargeted"
		}
		return artifactDisplayName(int32(v))
	case "artifact_name":
		return artifactDisplayName(int32(parseInt()))
	case "rarity":
		i := parseInt()
		if i >= 0 && i < len(rarityNames) {
			return rarityNames[i]
		}
		return rawVal
	case "tier":
		i := parseInt()
		if i >= 0 && i < len(tierNames) {
			return tierNames[i]
		}
		return rawVal
	case "spec_type":
		return rawVal
	case "time_bucket":
		return rawVal
	}
	return rawVal
}

File: src/EggLedger.Domain/Reports/Labels.cs
using System.Globalization;
using EggLedger.Domain.Ei;
using EggLedger.Domain.LedgerData;
using Ei;

namespace EggLedger.Domain.Reports;

/// <summary>
/// Raw-value to display-string conversion and sort ordering for report group-by dimensions.
/// Port of Go reports/labels.go.
/// </summary>
public static class Labels
{
    private static readonly string[] RarityNames = ["Common", "Rare", "Epic", "Legendary"];
    private static readonly string[] TierNames = ["T1", "T2", "T3", "T4"];

    private static LedgerDisplayData Config => LedgerData.LedgerData.Config;

    /// <summary>
    /// Display name for an artifact enum value via ledgerdata.ArtifactTargets,
    /// falling back to CasedName. Port of Go artifactDisplayName.
    /// </summary>
    private static string ArtifactDisplayName(int v)
    {
        var protoName = EnumNames.ProtoName((ArtifactSpec.Name)v);
        foreach (var t in Config.ArtifactTargets)
        {
            if (t.Name == protoName)
            {
                return t.DisplayName;
            }
        }
        return ((ArtifactSpec.Name)v).CasedName();
    }

    /// <summary>GROUP BY dimensions whose raw SQL value is an integer. Port of Go numericGroupBys.</summary>
    private static readonly HashSet<string> NumericGroupBys = new(StringComparer.Ordinal)
    {
        "ship_type", "duration_type", "level", "mission_type",
        "rarity", "tier", "artifact_name", "mission_target",
    };

    /// <summary>
    /// True when rawA sorts before rawB for the groupBy. Numeric dimensions compare
    /// by integer, others ordinal. Port of Go LabelSortLess.
    /// </summary>
    public static bool LabelSortLess(string groupBy, string rawA, string rawB)
    {
        if (NumericGroupBys.Contains(groupBy))
        {
            var okA = long.TryParse(rawA, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var a);
            var okB = long.TryParse(rawB, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var b);
            if (okA && okB)
            {
                return a < b;
            }
        }
        return string.CompareOrdinal(rawA, rawB) < 0;
    }

    /// <summary>Converts a raw SQL GROUP BY value to a display string. Port of Go FormatLabel.</summary>
    public static string FormatLabel(string groupBy, string rawVal)
    {
        int ParseInt()
        {
            int.TryParse(rawVal, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var v);
            return v;
        }

        switch (groupBy)
        {
            case "ship_type":
                return ((MissionInfo.Spaceship)ParseInt()).Name();
            case "duration_type":
                return ((MissionInfo.DurationType)ParseInt()).Display();
            case "level":
                return $"Level {rawVal}";
            case "mission_type":
                return ((MissionInfo.MissionType)ParseInt()).Display();
            case "mission_target":
                var v = ParseInt();
                if (v < 0)
                {
                    return "None (Pre 1.27)";
                }
                if (v == 0)
                {
                    return "Untargeted";
                }
                return ArtifactDisplayName(v);
            case "artifact_name":
                return ArtifactDisplayName(ParseInt());
            case "rarity":
                var ri = ParseInt();
                if (ri >= 0 && ri < RarityNames.Length)
                {
                    return RarityNames[ri];
                }
                return rawVal;
            case "tier":
                var ti = ParseInt();
                if (ti >= 0 && ti < TierNames.Length)
                {
                    return TierNames[ti];
                }
                return rawVal;
            case "spec_type":
                return rawVal;
            case "time_bucket":
                return rawVal;
        }
        return rawVal;
    }
}
